jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -2368,21 +2377,107 @@ private ClassDefiner classDefiner(byte[] bytes, ClassOption... classOptions) thr
 				throw new IllegalAccessException();
 			}
 
+			int flags = 0;
+			for (ClassOption opt : classOptions) {
+				flags |= opt.toFlag();
+			}
+			String targetClassName = getClassNameAndCheckClassBytes(bytes);
+			return makeHiddenClassDefiner(targetClassName, bytes, flags);
+			
+		}
+		
+		private static int readShortFromByteArray(byte[] bytes, int offset) {
+			if ((offset + 2) <= bytes.length) {
+				return Short.toUnsignedInt(ByteBuffer.wrap(bytes).getShort(offset));
+			} else {
+				throw new ArrayIndexOutOfBoundsException();
+			}
+		}
+		
+		private static int readIntFromByteArray(byte[] bytes, int offset) {
+			if ((offset + 4) <= bytes.length) {
+				return ByteBuffer.wrap(bytes).getInt(offset);
+			} else {
+				throw new ArrayIndexOutOfBoundsException();
+			}
+		}
+		private String getClassNameAndCheckClassBytes(byte[] bytes) {
```

Comment:
Why is this being done here rather than in cfreader?  Do we not have the context there?  Could we get it by passing another flag thru?

Reply:
The checks added here (and their order) are all forced by hidden class tests, they expect certain type of exception/error. I think ClassReader is doing checks differently (or maybe in different order), using ClassReader results a different type of exception/error, which makes those tests fail.

Reply:
I would rather this logic stay in the native `cfreader.c` implementation then duplicate it here as well.  Even if that means making additional native calls to do the processing with cfreader.

Reply:
Ahh, I see. You mean cfreader.c

Reply:
We may move the magicNum and version checks into cfreader.c, and call the native from here. I guess other checks like the descriptor of this class and whether this class is in the same package as the host class can still be here in the java level. All these checks are in java level here in RI.

Reply:
Moved the magic number and class version checks into cfreader.c.